2001-03-14  Jakub Jelinek  <jakub@redhat.com>

	* reg-stack.c (stack_result): Unconditionally use
	FUNCTION_OUTGOING_VALUE resp. FUNCTION_VALUE.

	* g++.old-deja/g++.other/regstack.C: New test.

--- gcc/reg-stack.c.jj	Mon Mar 12 11:45:08 2001
+++ gcc/reg-stack.c	Wed Mar 14 18:11:59 2001
@@ -786,11 +786,7 @@ stack_result (decl)
     return 0;
 
   result = DECL_RTL (DECL_RESULT (decl));
-  /* ?!?  What is this code supposed to do?  Can this code actually
-     trigger if we kick out aggregates above?  */
-  if (result != 0
-      && ! (GET_CODE (result) == REG
-	    && REGNO (result) < FIRST_PSEUDO_REGISTER))
+  if (result != 0)
     {
 #ifdef FUNCTION_OUTGOING_VALUE
       result
--- gcc/testsuite/g++.old-deja/g++.other/regstack.C.jj	Wed Mar 14 18:10:24 2001
+++ gcc/testsuite/g++.old-deja/g++.other/regstack.C	Wed Mar 14 18:10:06 2001
@@ -0,0 +1,31 @@
+// Special g++ Options: -O2
+// Skip if not target: i?86-*-*
+
+inline double foo (double x)
+{
+  register double r;
+  asm volatile ("fsqrt" : "=t" (r) : "0" (x));
+  return r;
+}
+
+struct X
+{
+  ~X() {}
+};
+
+int b;
+
+double bar (X x)
+{
+  if (b)
+    return 1.0;
+  return 36.0 * foo (36.0);
+}
+
+int main ()
+{
+  X x;
+  if (bar (x) != 216.0)
+    return 1;
+  return 0;
+}
